lib/evmone/eof.cpp

```diff
@@ -90,87 +197,494 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+
+        successors.clear();
+
+        // immediate_size for RJUMPV depends on the code. It's calculated below.
+        if (opcode != OP_RJUMP && !instr::traits[opcode].is_terminating && opcode != OP_RJUMPV)
+        {
+            auto next = i + instr::traits[opcode].immediate_size + 1;
+            if (next >= code.size())
+                return {EOFValidationError::no_terminating_instruction, -1};
+
+            successors.push_back(next);
+        }
+
+        if (opcode == OP_RJUMP || opcode == OP_RJUMPI)
+        {
+            auto target_rel_offset_hi = code[i + 1];
+            auto target_rel_offset_lo = code[i + 2];
+            auto target_rel_offset =
+                static_cast<int16_t>((target_rel_offset_hi << 8) | target_rel_offset_lo);
+            successors.push_back(
+                static_cast<size_t>(target_rel_offset + 3 + static_cast<int32_t>(i)));
+        }
+
+        if (opcode == OP_RJUMPV)
+        {
+            auto count = code[i + 1];
+
+            auto next = i + count * 2 + 2;
+            if (next >= code.size())
+                return {EOFValidationError::no_terminating_instruction, -1};
+
+            auto beg = stack_heights.begin() + static_cast<int32_t>(i) + 1;
+            std::fill_n(beg, count * 2 + 1, -2);
+
+            successors.push_back(next);
+
+            for (uint16_t k = 0; k < count; ++k)
+            {
+                auto target_rel_offset_hi = code[i + k * 2 + 2];
+                auto target_rel_offset_lo = code[i + k * 2 + 3];
+                auto target_rel_offset =
+                    static_cast<int16_t>((target_rel_offset_hi << 8) | target_rel_offset_lo);
+                successors.push_back(static_cast<size_t>(
+                    static_cast<int32_t>(i) + 2 * count + target_rel_offset + 2));
+            }
+        }
+        else
+        {
+            auto beg = stack_heights.begin() + static_cast<int32_t>(i) + 1;
+            std::fill_n(beg, instr::traits[opcode].immediate_size, -2);
+        }
+
+        stack_height += stack_height_change;
+
+        for (auto& s : successors)
+        {
+            if (stack_heights[s] == -1)
+            {
+                stack_heights[s] = stack_height;
+                worklist.push(s);
+            }
+            else if (stack_heights[s] != stack_height)
+                return {EOFValidationError::stack_height_mismatch, -1};
+        }
+
+        if (opcode == OP_RETF && stack_height != funcs_in_outs[func_index].outputs_num)
+            return {EOFValidationError::non_empty_stack_on_terminating_instruction, -1};
+    }
+
+    auto msh_it = std::max_element(stack_heights.begin(), stack_heights.end());
+
+    if (std::find(stack_heights.begin(), stack_heights.end(), -1) != stack_heights.end())
+        return {EOFValidationError::unreachable_instructions, -1};
+
+    return {EOFValidationError::success, *msh_it};
+}
+
+std::variant<EOF1Header, EOFValidationError> validate_eof1(  // NOLINT(misc-no-recursion)
```

Comment:
I counted already 7 places, including this, where we suppress recursion warning. Does it make sense to disable globally?

Reply:
Here I actually decided better to rewrite without recursion.

Reply:
> Here I actually decided better to rewrite without recursion.

I would do it in a separate PR #794